stl/inc/functional

```diff
@@ -1207,6 +1207,645 @@ _NODISCARD bool operator!=(nullptr_t, const function<_Fty>& _Other) noexcept {
+_NODISCARD _Rx __stdcall _Function_inv_large(const _Move_only_function_data& _Self, _Types&&... _Args) noexcept(_Noex) {
+    return _Invoker_ret<_Rx>::_Call(
+        static_cast<_VtInvQuals>(*_Self._Large_fn_ptr<_Vt>()), _STD forward<_Types>(_Args)...);
+}
+
+template <class _Vt>
+void __stdcall _Function_move_small(_Move_only_function_data& _Self, _Move_only_function_data& _Src) noexcept {
+    const auto _Src_fn_ptr = _Src._Small_fn_ptr<_Vt>();
+    ::new (_Self._Buf_ptr<_Vt>()) _Vt(_STD move(*_Src_fn_ptr));
+    _Src_fn_ptr->~_Vt();
+    _Self._Impl = _Src._Impl;
+}
+
+template <size_t _Size>
+void __stdcall _Function_move_memcpy(_Move_only_function_data& _Self, _Move_only_function_data& _Src) noexcept {
+    _CSTD memcpy(&_Self._Data, &_Src._Data, _Size); // Copy Impl* and functor data
+}
+
+inline void __stdcall _Function_move_large(_Move_only_function_data& _Self, _Move_only_function_data& _Src) noexcept {
+    _CSTD memcpy(&_Self._Data, &_Src._Data, _Minimum_function_size); // Copy Impl* and functor data
+}
+
+template <class _Vt>
+void __stdcall _Function_destroy_small(_Move_only_function_data& _Self) noexcept {
+    _Self._Small_fn_ptr<_Vt>()->~_Vt();
+}
+
+inline void __stdcall _Function_deallocate_large_default_aligned(_Move_only_function_data& _Self) noexcept {
+    ::operator delete(_Self._Large_fn_ptr<void>());
+}
+
+template <size_t _Align>
+void __stdcall _Function_deallocate_large_overaligned(_Move_only_function_data& _Self) noexcept {
+    ::operator delete (_Self._Large_fn_ptr<void>(), align_val_t{_Align});
+}
+
+template <class _Vt>
+void __stdcall _Function_destroy_large(_Move_only_function_data& _Self) noexcept {
+    const auto _Pfn = _Self._Large_fn_ptr<_Vt>();
+    _Pfn->~_Vt();
+    if constexpr (alignof(_Vt) <= __STDCPP_DEFAULT_NEW_ALIGNMENT__) {
+        ::operator delete(static_cast<void*>(_Pfn));
+    } else {
+        ::operator delete (static_cast<void*>(_Pfn), align_val_t{alignof(_Vt)});
+    }
+}
+
+template <class _Vt>
+inline constexpr size_t _Function_small_copy_size = // We copy Impl* and the functor data at once
+    _Move_only_function_data::_Buf_offset<_Vt> + // Impl* plus possible alignment
+    (size_t{sizeof(_Vt) + sizeof(void*) - 1} & ~size_t{sizeof(void*) - 1}); // size in whole pointers
+
+template <class _Vt, class... _CTypes>
+_NODISCARD void* _Function_new_large(_CTypes&&... _Args) {
+    struct _NODISCARD _Guard_type {
+        void* _Ptr;
+
+        ~_Guard_type() {
+            // _Ptr is not nullptr only if an exception is thrown as a result of _Vt construction.
+            // Check _Ptr before calling operator delete to save the call in the common case.
+            if (_Ptr) {
+                if constexpr (alignof(_Vt) <= __STDCPP_DEFAULT_NEW_ALIGNMENT__) {
+                    ::operator delete(_Ptr);
+                } else {
+                    ::operator delete (_Ptr, align_val_t{alignof(_Vt)});
+                }
+            }
+        }
+    };
+
+    _Guard_type _Guard;

```

Comment:
Excellent catch, thanks!
I did not want to initialize `_Guard_type` with `nullptr`, I instead moved it after `::operator new`. I think this way codegen may be better in case if the functor constructor does not throw.

Reply:
My understanding is that the optimizer is good at eliminating dead stores. Since uninitialized variables are scary, I'd prefer to see a data member initializer to `nullptr`, although because this guard is completely local, I don't think it's absolutely necessary.

Reply:
Well, I see that on x64 the compiler does not optimize away dead store, but still the effect is small https://godbolt.org/z/qnhehMx9Y

On x86 however where SEH is not free, the cost is easily seen: https://godbolt.org/z/j6EzqdG5c

Reply:
Note that currently the data member is initialized by brace-init. Can add constructor if you wish, can also add initialization of non-member `_Ptr`, this all will not impact the optimization. (I find it hard to explain my compiler intuition, but I checked this statement is correct).

Reply:
Ok, no further change requested. I'll go ahead and resolve this - thanks again @muellerj2, we really appreciate how you caught this bug! :heart_eyes_cat: